File: Tree/104. Maximum Depth of Binary Tree/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# First Try
# class Solution:
#     def maxDepth(self, root):
#         def dfs(node, depth):
#             if not node: return depth
#             depth += 1
#             depth_l = dfs(node.left, depth)
#             depth_r = dfs(node.right, depth)
#             return max(depth_l, depth_r)

#         return dfs(root, 0)

# This solution is more clear
class Solution:
    def maxDepth(self, root):
        if not root: return 0
        depth_l = self.maxDepth(root.left)
        depth_r = self.maxDepth(root.right)
        return max(depth_l, depth_r) + 1

File: Tree/104. Maximum Depth of Binary Tree/go/main.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// DFS solution
func maxDepth(root *TreeNode) int {
	return dfs(root)
}

func dfs(root *TreeNode) int {
	if root == nil {
		return 0
	}

	return max(dfs(root.Left), dfs(root.Right)) + 1
}

func max(i, j int) int {
	if i > j {
		return i
	}
	return j
}

// BFS solution
func maxDepthBFS(root *TreeNode) int {
	return bfs(root)
}

func bfs(root *TreeNode) int {
	if root == nil {
		return 0
	}

	q := []*TreeNode{}
	depth := 0

	q = append(q, root)

	for len(q) != 0 {
		size := len(q)
		// pop the nodes in current level
		for i := 0; i < size; i++ {
			node := q[0]
			q = q[1:]
			// push child nodes
			if node.Left != nil {
				q = append(q, node.Left)
			}
			if node.Right != nil {
				q = append(q, node.Right)
			}
		}
		depth++
	}

	return depth
}
